### PR TITLE
[chore] small tweaks & bump VMs in CI

### PR DESCRIPTION
**Pull request by kelset**
#### Please select one of the following
- [ ] I am removing an existing difference between facebook/react-native and microsoft/react-native-macos :thumbsup:
- [ ] I am cherry-picking a change from Facebook's react-native into microsoft/react-native-macos :thumbsup:
- [ ] I am making a fix / change for the macOS implementation of react-native
- [x] I am making a change required for Microsoft usage of react-native

*(wasn't really sure what to pick tbh)*

## Summary

While digging around the code I noticed that the ADO vms used by this repo are a bit outdated. So I tweaked a bit the vars.yml file, added all the referenced VMs, and updated to newer versions.
I've matched the "latest" ones but used the fixed reference, according to the docs: https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml#software
This is aligned with what react-native-test-app is using: https://github.com/microsoft/react-native-test-app/blob/trunk/.github/workflows/build.yml

Moreover, I've set the Xcode version to match the one set as default by the new macos VM: https://github.com/actions/virtual-environments/blob/main/images/macos/macos-11-Readme.md#xcode

This is actually something we might want to discuss about, since technically in the CircleCI config we should also change to match, potentially: https://github.com/microsoft/react-native-macos/blob/main/.circleci/config.yml#L55

I didn't change it since it seems to be a config we have in place just because we're a fork of core and they have a CircleCI config? https://github.com/facebook/react-native/blob/1b31d6bb582768ccbe92d3c1a9e43354a8c531e5/.circleci/config.yml#L55

So wanted to check before diverging from upstream.

To sum it up:
* `ubuntu-18.04` -> `ubuntu-20.04`
* `macOS-10.15` -> `macOS-11`
* `Xcode_12_4` -> `Xcode_13_1`

## Test Plan

Running on CI is itself the test, let's see how it goes :)

### CODE REVIEW

File: .ado/publish.yml [hunk 1]
```diff
@@ -107,7 +107,7 @@ jobs:
   - job: RNMacOSInitNpmJSPublish
     displayName: react-native-macos-init Publish to npmjs.org
     pool:
-      vmImage: vs2017-win2016
+      vmImage: $(VmImageWindows)
```
Comment (Saadnajmi) [inline .ado/publish.yml:110]: I do a similar change in https://github.com/microsoft/react-native-macos/pull/942 , so there might be a merge conflict pending which change goes in first (probably yours)

File: .ado/variables/vars.yml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+variables:
+  VmImageApple: macOS-11
+  VmImageWindows: windows-2019
+  VmImageUbuntu: ubuntu-20.04
+  slice_name: 'Xcode_13_1'
+  xcode_version: '/Applications/Xcode_13.1.app'
```
Comment (Saadnajmi) [inline .ado/variables/vars.yml:6]: @HeyImChris was previously working on getting this repo on Xcode 13, and I think ran into issues? He'll probably want to comment. We care a lot about keeping Xcode version used in react-native-macos in sync with downstream, so this would be great!

File: .ado/publish.yml (hunk 1 above)
Comment (kelset) [reply .ado/publish.yml:110]: yeah makes sense - lmk how that goes 👍

File: .ado/variables/vars.yml (hunk 2 above)
Comment (kelset) [reply .ado/variables/vars.yml:6]: I've used Xcode 13 without issues against this repo for the RNtester, which is why I did this bump - what were the problems hit?

File: .ado/publish.yml (hunk 1 above)
Comment (Saadnajmi) [reply .ado/publish.yml:110]: I just merged my change. We should no longer have any windows VMs in our ADO pipeline =)

File: .ado/variables/vars.yml (hunk 2 above)
Comment (Saadnajmi) [reply .ado/variables/vars.yml:6]: I think it was mostly with downstream integration into SDX-Platform and the like. I'm not too sure.

Comment (HeyImChris) [reply .ado/variables/vars.yml:5]: I think we're good to go forward with Xcode 13 and Big Sur support now! I created a PR today doing just this but then Saad reminded me of your review. Let's just get this in soon so our CI is up to date - @kelset let me know if you need help with the CI or anything else. It can be a bit flaky at times

**Comment by kelset** [comment]
Update on this PR - I spent some time trying to get to the bottom of the CI ❌

* The android one I think is still happening because the fix by @amgleitman has been merged only against 0.66 branch: https://github.com/microsoft/react-native-macos/pull/989
* The iOS ones seem to be actually something to look more into, as I can repro locally:
Screenshot 2022-01-27 at 15 00 36
I did some basic digging around but I couldn't find anything specific that could address it, so I'll go into the 0.66 branch, verify if it happens there, and adapt accordingly. Will report back afterwards.

**Review by HeyImChris — APPROVED** [review]
